pkg/proxy/proxy.go

```diff
@@ -1,267 +0,0 @@
-	"github.com/brancz/kube-rbac-proxy/pkg/authz"
-	"k8s.io/apiserver/pkg/authentication/authenticator"
-	"k8s.io/apiserver/pkg/authentication/user"
-	"k8s.io/apiserver/pkg/authorization/authorizer"
-	clientset "k8s.io/client-go/kubernetes"
-	"k8s.io/klog/v2"
-)
-
-// Config holds proxy authorization and authentication settings
-type Config struct {
-	Authentication *authn.AuthnConfig
-	Authorization  *authz.Config
-}
-
-type kubeRBACProxy struct {
-	// authenticator identifies the user for requests to kube-rbac-proxy
-	authenticator.Request
-	// authorizer determines whether a given authorization.Attributes is allowed
-	authorizer.Authorizer
-	// authorizerAttributesGetter implements retrieving authorization attributes for a respective request.
-	authorizerAttributesGetter *krpAuthorizerAttributesGetter
-	// config for kube-rbac-proxy
-	Config Config
-}
-
-func new(authenticator authenticator.Request, authorizer authorizer.Authorizer, config Config) *kubeRBACProxy {
-	return &kubeRBACProxy{authenticator, authorizer, newKubeRBACProxyAuthorizerAttributesGetter(config.Authorization), config}
-}
-
-// New creates an authenticator, an authorizer, and a matching authorizer attributes getter compatible with the kube-rbac-proxy
-func New(client clientset.Interface, config Config, authorizer authorizer.Authorizer, authenticator authenticator.Request) (*kubeRBACProxy, error) {
-	return new(authenticator, authorizer, config), nil
-}
-
-// Handle authenticates the client and authorizes the request.
-// If the authn fails, a 401 error is returned. If the authz fails, a 403 error is returned
-func (h *kubeRBACProxy) Handle(w http.ResponseWriter, req *http.Request) bool {
-	ctx := req.Context()
-	if len(h.Config.Authentication.Token.Audiences) > 0 {
-		ctx = authenticator.WithAudiences(ctx, h.Config.Authentication.Token.Audiences)
-		req = req.WithContext(ctx)
-	}
-
-	// Authenticate
-	u, ok, err := h.AuthenticateRequest(req)
-	if err != nil {
-		klog.Errorf("Unable to authenticate the request due to an error: %v", err)
-		http.Error(w, "Unauthorized", http.StatusUnauthorized)
-		return false
-	}
-	if !ok {
-		http.Error(w, "Unauthorized", http.StatusUnauthorized)
-		return false
-	}
-
-	// Get authorization attributes
-	allAttrs := h.authorizerAttributesGetter.GetRequestAttributes(u.User, req)
-	if len(allAttrs) == 0 {
-		msg := "Bad Request. The request or configuration is malformed."
-		klog.V(2).Info(msg)
-		http.Error(w, msg, http.StatusBadRequest)
-		return false
-	}
-
-	for _, attrs := range allAttrs {
-		// Authorize
-		authorized, reason, err := h.Authorize(ctx, attrs)
-		if err != nil {
-			msg := fmt.Sprintf("Authorization error (user=%s, verb=%s, resource=%s, subresource=%s)", u.User.GetName(), attrs.GetVerb(), attrs.GetResource(), attrs.GetSubresource())
-			klog.Errorf("%s: %s", msg, err)
-			http.Error(w, msg, http.StatusInternalServerError)
-			return false
-		}
-		if authorized != authorizer.DecisionAllow {
```

Comment:
This changes no-opinion to false instead of falling through, making the behavior dependent on the ordering of headers in a logical map, right?

Reply:
We are iterating over all elements of `allAttrs` and all elements need to be authorized. This is `AND` logic. So the ordering doesn't matter.